### PR TITLE
Fix ReactServerAgent CORS preflight

### LINKED ISSUE #619: ReactServerAgent simple GET request forces CORS preflight
Make a simple GET request via ReactServerAgent and CORS will kick off a OPTIONS preflight.   Make the exact same request directly via SuperAgent and it will work fine.   Best I can tell is that ReactServerAgent is setting Content-Type on the GET request.   Perhaps here?
`
if (this._type && this._type !== 'form-data') {
        req.type(this._type);
}
`

This may be as simply as adding `&& this._method !== 'GET'` but it could be somewhere else as well.

### PR DESCRIPTION
**Pull request by drewpc**
Fixes #619.  ReactServerAgent shouldn't create a CORS preflight request when executing simple GET queries of an external API.  This patch fixes that problem as well as an unidentified problem where a HEAD request generate a CORS preflight request as well, because it was attempting to pass data to the request using SuperAgent's `.send()` method instead of `.query()`.

### CODE REVIEW

File: packages/react-server/core/ReactServerAgent/Request.js [hunk 1]
```diff
@@ -206,34 +206,48 @@ function buildSuperagentRequest() {
...
+				req.send(postParams);
 			}
-			postParams = formData;
-		} else {
-			throw new Error(`ReactServerAgent.type("form-data") not allowed server-side`);
-		}
-	}
 
-	req.send(postParams);
+			break;
```
Comment (doug-wade) [inline packages/react-server/core/ReactServerAgent/Request.js]: I'm being a little stickler meeseeks

but what about the other http methods (CONNECT, OPTIONS, TRACE)?  I feel like we should at least have a default case to emit a warning or error or something.

Comment (drewpc) [inline packages/react-server/core/ReactServerAgent/Request.js]: Fair point.  We don't have functions for those methods, so I didn't implement a check for those.  I can make this more generic.

**Comment by roblg** [comment]
FYI, I'd like to take a closer look at this one. I'll try to get to it this morning.

File: packages/react-server/core/ReactServerAgent/Request.js [hunk 2]
```diff
@@ -233,7 +228,35 @@ function buildSuperagentRequest() {
 		}
 	}
 
-	req.send(postParams);
+	let hasQueryParams = (this._queryParams.length > 0),
+		hasPostParams = (Object.keys(this._postParams).length > 0);
```
Comment (roblg) [inline packages/react-server/core/ReactServerAgent/Request.js]: Nit: I think these can be `const`?

File: packages/react-server/core/ReactServerAgent/Request.js [hunk 3]
```diff
@@ -233,7 +228,35 @@ function buildSuperagentRequest() {
...
+
+	if (hasPostParams) {
+		// superagent has some weird, implicit file upload support
+		// that only works if you don't set `type`.
+		if (this._type && this._type !== 'form-data') {
+			req.type(this._type);
+		}
+
+		req.send(postParams);
+	}
```
Comment (roblg) [inline packages/react-server/core/ReactServerAgent/Request.js]: I think there might be an edge case here. It seems like one should be able to POST an empty JSON object to an endpoint, and I think we can't anymore. (Though to be fair, I think that may have been the default before, which probably also doesn't make sense?)

I suspect that there have always been lots of edge cases here that we just haven't run into yet. (I just realized that superagent accepts strings _and_ objects for `send`, and we only accept objects.

**Comment by roblg** [comment]
Gah. I clicked 'approve' went I meant to just leave the review comments.

File: packages/react-server/core/ReactServerAgent/Request.js (hunk 3 above)
Comment (drewpc) [inline packages/react-server/core/ReactServerAgent/Request.js]: Correct...any thoughts on the right way to handle this?  Here's the basic situation: on GET/HEAD requests, if we set `type` and/or `send` to any value (even empty values), the GET/HEAD requests will send a CORS preflight.

Comment (drewpc) [inline packages/react-server/core/ReactServerAgent/Request.js]: @roblg I handled the edge case were someone wants to send an empty JSON object now.  And there's a test for it!

Comment (roblg) [inline packages/react-server/core/ReactServerAgent/Request.js]: Nice and nice!

File: packages/react-server/core/ReactServerAgent/Request.js [hunk 4]
```diff
@@ -233,7 +236,37 @@ function buildSuperagentRequest() {
...
+			}
+			break;
+
+		case 'POST':
+		case 'PATCH':
+		case 'PUT':
+			if (hasQueryParams) {
+				logger.info(`Attempting a ${this._method} request with query data (using SuperAgent's .query() function). This might not be what you want.`);
+			}
+			break;
```
Comment (doug-wade) [inline packages/react-server/core/ReactServerAgent/Request.js]: should we have a warning or error or something for the default case?

File: packages/react-server/core/ReactServerAgent/Request.js [hunk 5]
```diff
@@ -71,9 +74,14 @@ Request.prototype.query = function (queryParams) {
 
 Request.prototype.send = function (postParams) {
 	if (typeof postParams === 'undefined') {
-		return merge({}, this._postParams);
+		return merge({}, this._postParams || {});
+	}
+	if (postParams !== null) {
```
Comment (doug-wade) [inline packages/react-server/core/ReactServerAgent/Request.js]: I always get a little skittish when `undefined` and `null` are treated differently; is there some way we can consolidate them?

Comment (drewpc) [inline packages/react-server/core/ReactServerAgent/Request.js]: I 100% agree.  However, it seems that the current implementation of all of these functions does something different if you call it like `.send()` (returns the currently set post parameters) vs `.send({})` (which saves the post parameters and returns `this`).  In my opinion, that is a bad design for all of those functions...trying to combine setters + getters.  That's outside the scope of this PR though.

File: packages/react-server/core/ReactServerAgent/Request.js (hunk 4 above)
Comment (drewpc) [inline packages/react-server/core/ReactServerAgent/Request.js]: I don't think so.  If someone is using a method other than GET, HEAD, POST, PUT, or PATCH and they're setting query/post parameters, then they probably know what they are doing and aren't worried about a CORS preflight request being emitted.

File: packages/react-server/core/ReactServerAgent/Request.js (hunk 5 above)
Comment (drewpc) [inline packages/react-server/core/ReactServerAgent/Request.js]: FWIW, looking at the corresponding SuperAgent functions, they _always_ return `this`.  They are not trying to combine a setter with a getter function.

Comment (roblg) [inline packages/react-server/core/ReactServerAgent/Request.js]: FWIW, I have no recollection of why I thought it was a good idea to combine set/get. I think I was probably thinking of properties, or looking at something else. I doubt they need to be getters as well, but I haven't poked around to see what would need to be updated.
